### DISCUSSION TITLE
Understanding the Depth to Space Layers of the CNN Upscalers

### DISCUSSION THREAD

**Discussion by sb2702** [bloc97/Anime4K#204] [Q&A] [answered]
Hi! 👋

I'm in the very beginning stages of building [a web library / sdk ](https://github.com/sb2702/websr/tree/main)to make it super easy to run super resolution in the browser. I'm not here to share the project (yet), it's still very much in construction.

As a starting point / proof of concept, I'd like to port a few Anime4K shaders to WebGPU, especially the upscaler CNN networks, since they are conceptually very easy to write shaders for. 

I implemented the [Anime4K_Upscale_CNN_x2_S.glsl ](Anime4K_Upscale_CNN_x2_S.glsl) network [here](https://github.com/sb2702/websr/blob/main/src/networks/anime4k/cnn-2x-s.ts) and I think it works (see output below)?

<img width="512" alt="Screen Shot 2023-09-08 at 6 25 44 PM" src="https://github.com/bloc97/Anime4K/assets/5678502/d0e462dc-b75b-4b00-8593-efadcce44ed5">

I just wanted to confirm I'm understanding the depth to space layer correctly, since while I'm familiar with WebGL, I've never used MPV's hooks and am not where what some of [these variable values](https://github.com/bloc97/Anime4K/blob/8e39551ce96ed172605c89b7dd8be855b5502cc9/glsl/Upscale/Anime4K_Upscale_CNN_x2_S.glsl#L151) are at runtime.

I understand that you have a `conv2d_last_tf` texture which is `w`x`h` with 4 channels, and the depth to space layer converts that to a `(w*2)x(h*2)` 1-channel residual, which then gets added to every channel of the bicubic upscaled input image.

By trial and error, I found out that the residual is constructed by organizing the `conv2d_last_tf` values in the following order:

![equation (1)](https://github.com/bloc97/Anime4K/assets/5678502/7f0f612f-bc7c-45e5-b428-2519b4bbb45f)

The actual WGSL langauge (which is similar enough to OpenGL) I used is:

    let x_floor  = u32(fract(input.tex_coord.x*256.0)*2.0);
                    
    let y_floor  = u32(fract(input.tex_coord.y*256.0)*2.0);
    
    //I don t know, I think this is right? I found this by trial and error
    let c_index: u32 = 2*x_floor + 1-y_floor;  

    let x = i32(256.0*(input.tex_coord.x));
    let y = i32(256.0*(input.tex_coord.y));
    
    let value = textureLoad(featureMap, vec2<i32>(x, y), 0)[c_index];

Where `input.tex_coord.xy`  is a `vec2` that ranges between 0 and 1 along both the x and y directions, and is run on a fragment shader on the full-sized `(w*2)x(h*2)` output.

I suspect one issue might be the fact that in WebGPU you need to flip the Y direction before rendering to canvas, but in either case, the order of the pixel shuffle seems weird, and this line `[i0.y * 2 + i0.x]` [here](https://github.com/bloc97/Anime4K/blob/master/glsl/Upscale/Anime4K_Upscale_CNN_x2_S.glsl#L153) makes me think it should be the a pixel shuffle order of 

![equation (2)](https://github.com/bloc97/Anime4K/assets/5678502/e5c7a682-fce6-417d-b615-a53f1df9d4f9)

Which is what I would have expected, but doing so seems to result in a residual map that seems just wrong?

<img width="512" alt="Screen Shot 2023-09-08 at 6 56 17 PM" src="https://github.com/bloc97/Anime4K/assets/5678502/9cde682f-bdfb-46dd-b171-6e2622f07217">

For the time being I'll just stick with the order that seems to be outputting the correct residual map, but this did seem confusing so I figured I'd ask to clarify if anyone has a clue as to if I'm doing something fundamentally wrong.

**Comment by bloc97** [comment]
That's a really cool concept! Would love to see upscaling for web content...
You got the understanding of the last layer correctly, it is just a pixel shuffle layer. However, looking at your examples, both results look wrong (they all have artifacts), maybe it is something else that has a bug? Shaders for WebGL and OpenGL should be applied in the same orientation right?

**Reply by sb2702** [reply]
First of all, thank you so much!

There's probably something else going on then under the hood, I'll have to debug. I don't suppose you, by any chance, have the model checkpoints for any of those glsl files?

If not, no worries. I'll probably first need to start with a network I can play around with in Tensorflow ([this notebook](https://github.com/bloc97/Anime4K/blob/master/tensorflow/Train_Model.ipynb) seems like a good starting point, I can just train a fresh network on the Synla+ data set?) so I can get the intermediate feature maps from inference on a reference image and debug layer by layer, to see where my webgpu pipeline goes wrong.

That would also have the double-benefit of bench-marking performance of using TFJS vs hard-coded shaders. I've been assuming TFJS introduces too much overhead for real-time web video, but having good numbers around that would help.

**Reply by bloc97** [reply, marked as answer]
That would be a good start, training on a subset of the synla dataset using the provided demo notebook takes only a few minutes for the smallest network. (Even if the final networks were not trained exclusively on the synla dataset, for testing purposes it is enough)

**Reply by sb2702** [reply]
Thank you, will do! Hopefully I can debug and figure out what's going on, and either way, having automated tests to know the output is correct and matches the tensorflow output seems like a good idea anyway. Once I get it working for the debug weights, for the production network weights, hopefully I can just run the production glsl files through mpv and verify the final output matches what is coming out from WebGPU. Thanks again!

**Reply by sb2702** [reply]
Just closing the loop on this, it turns out WebGPU does flip the Y axis, and that was flipping all the feature maps upside down,  so of course there were artifacts in the final image.

<img width="1034" alt="Screen Shot 2023-09-10 at 8 53 33 PM" src="https://github.com/bloc97/Anime4K/assets/5678502/64980777-8b0f-4a77-9177-55acf4b85980">

12 hours of coding to extract and compare the feature maps and figure out the issue, and literally 1 second to fix it 😅

I at least feel better now having built a [test suite](https://sb2702-files.s3.amazonaws.com/ml/datasets/websr-inference-test/y_flip/index.html). The tensorflow notebook was super helpful for this! Thanks again!
